File: qbopt/hir/dump.py
"""Stable semantic projection used to diff source and object frontends."""

from qbopt.model import mir
from qbopt.model import memory
from qbopt.hir.lower import Lowered


class _Names:
    def __init__(self, body: mir.MirBody) -> None:
        self.blocks = {block.at: number for number, block in enumerate(body.blocks, 1)}
        self.values: dict[mir.Value, int] = {}
        self.objects: dict[memory.Object, int] = {}

    def value(self, value: mir.Value) -> str:
        number = self.values.setdefault(value, len(self.values) + 1)
        return f"f{number}" if value.flags else f"v{number}"

    def block(self, at: int) -> str:
        return f"b{self.blocks[at]}"

    def object(self, object_: memory.Object) -> str:
        number = self.objects.setdefault(object_, len(self.objects) + 1)
        extent = "?" if object_.extent is None else str(object_.extent)
        return f"{object_.kind.value}{number}:{extent}"


def _arg(one: mir.Arg, names: _Names) -> str:
    match one:
        case mir.Held(value, width):
            return f"{names.value(value)}:{width}"
        case mir.Const(number, width):
            return f"{number}:{width}"
        case mir.Cell(ref):
            if ref.provenance is None:
                where = str(ref.where)
            else:
                where = "|".join(
                    f"{names.object(part.object)}[{part.low}:{part.high}:{part.stride}/{part.width}]"
                    for part in sorted(
                        ref.provenance.slices,
                        key=lambda item: (item.object.kind, repr(item.object.identity), item.low, item.high),
                    )
                )
            base = f"+{names.value(ref.base)}" if ref.base is not None else ""
            segment = f"@{names.value(ref.segment)}" if ref.segment is not None else ""
            return f"cell({where}{base}{segment}):{ref.width}"
        case mir.Symbol(space, index, offset, width, addend):
            displacement = offset + addend
            suffix = f"+{displacement}" if displacement > 0 else (str(displacement) if displacement else "")
            return f"{space.value}{index}{suffix}:{width}"
        case mir.FrameAddress(offset, width):
            sign = f"+{offset}" if offset > 0 else str(offset)
            return f"frame[{sign}]:{width}"
        case mir.Opaque(name=name):
            return name or "opaque"
        case _:
            return type(one).__name__.lower()


_INFIX = frozenset(
    {
        mir.Kind.ADD,
        mir.Kind.SUB,
        mir.Kind.ADD_CARRY,
        mir.Kind.SUB_BORROW,
        mir.Kind.MUL,
        mir.Kind.SMULHI,
        mir.Kind.DIV,
        mir.Kind.REM,
        mir.Kind.DIVMOD,
        mir.Kind.UDIVMOD,
        mir.Kind.AND,
        mir.Kind.OR,
        mir.Kind.XOR,
        mir.Kind.SHL,
        mir.Kind.SHR,
        mir.Kind.SAR,
        mir.Kind.LT,
        mir.Kind.LE,
        mir.Kind.GT,
        mir.Kind.GE,
        mir.Kind.EQ,
        mir.Kind.NE,
        mir.Kind.BELOW,
        mir.Kind.BELOW_EQ,
        mir.Kind.ABOVE,
        mir.Kind.ABOVE_EQ,
        mir.Kind.FADD,
        mir.Kind.FSUB,
        mir.Kind.FMUL,
        mir.Kind.FDIV,
        mir.Kind.FCOMPARE,
    }
)


def _operation(op: mir.Op, names: _Names) -> str:
    args = tuple(_arg(one, names) for one in op.args)
    results = tuple(names.value(one.value) if isinstance(one, mir.Held) else _arg(one, names) for one in op.results)
    defined = tuple(names.value(one) for one in op.defines)
    left = results or defined
    spelling = op.name if op.op.value == "funary" and op.name else op.kind.value

    if op.kind is mir.Kind.STORE and len(results) == 1 and len(args) == 1:
        return f"{results[0]} <- {args[0]}"
    if op.kind in _INFIX and len(args) == 2:
        expression = f"{args[0]} {spelling} {args[1]}"
    elif op.kind is mir.Kind.CALL:
        expression = f"call {op.name}({', '.join(args)})"
    elif op.kind is mir.Kind.BRANCH:
        expression = "branch" + (f" {args[0]}" if args else "")
    elif op.kind is mir.Kind.RETURN:
        expression = "return" + (f" {', '.join(args)}" if args else "")
    elif len(args) == 1:
        expression = f"{spelling} {args[0]}"
    elif not args:
        expression = spelling
    else:
        expression = f"{spelling}({', '.join(args)})"
    return f"{', '.join(left)} <- {expression}" if left else expression


def mir_text(lowered: Lowered) -> str:
    """Print meaning with incidental block, value, and object identities removed."""
    names = _Names(lowered.body)
    lines = [f"function {lowered.name} entry {names.block(lowered.body.entry)}"]
    for block in lowered.body.blocks:
        successors = ", ".join(names.block(one) for one in block.succ) or "-"
        lines.append(f"{names.block(block.at)} -> {successors}")
        for op in block.ops:
            target = f" -> {names.block(op.target)}" if op.target is not None else ""
            cases = "" if not op.cases else " " + repr(tuple((value, names.block(at)) for value, at in op.cases))
            floating = ""
            if op.floating is not None:
                inputs = ",".join(one.value for one in op.floating.inputs)
                floating = (
                    f" [{inputs}->{op.floating.result.value};"
                    f"{op.floating.precision.value}/{op.floating.rounding.value}]"
                )
            lines.append(f"  {_operation(op, names)}{target}{cases}{floating}")
    return "\n".join(lines) + "\n"

File: src/hir/dump.rs
//! Port of `qbopt/hir/dump.py`: stable semantic projection used to diff
//! source and object frontends.

use indexmap::IndexMap;

use crate::hir::lower::Lowered;
use crate::model::ir::Operation;
use crate::model::memory::MemoryObject;
use crate::model::mir::{self, Arg, OpCode};
use crate::support::pyrepr::Repr;

struct _Names {
    blocks: IndexMap<i64, usize>,
    values: IndexMap<mir::Value, usize>,
    objects: IndexMap<MemoryObject, usize>,
}

impl _Names {
    fn new(body: &mir::MirBody) -> Self {
        Self {
            blocks: body.blocks.iter().enumerate().map(|(number, block)| (block.at, number + 1)).collect(),
            values: IndexMap::new(),
            objects: IndexMap::new(),
        }
    }

    fn value(&mut self, value: mir::Value) -> String {
        let next = self.values.len() + 1;
        let number = *self.values.entry(value).or_insert(next);
        if value.flags { format!("f{number}") } else { format!("v{number}") }
    }

    fn block(&self, at: i64) -> String {
        format!("b{}", self.blocks[&at])
    }

    fn object(&mut self, object_: &MemoryObject) -> String {
        let next = self.objects.len() + 1;
        let number = *self.objects.entry(object_.clone()).or_insert(next);
        let extent = object_.extent.map_or_else(|| "?".to_owned(), |extent| extent.to_string());
        format!("{}{number}:{extent}", object_.kind.as_str())
    }
}

fn _arg(one: &Arg, names: &mut _Names) -> String {
    match one {
        Arg::Held(held) => format!("{}:{}", names.value(held.value), held.width),
        Arg::Const(constant) => format!("{}:{}", constant.n, constant.width),
        Arg::Cell(cell) => {
            let reference = &cell.r#ref;
            let where_ = match &reference.provenance {
                None => reference.where_().map_or_else(|| "None".to_owned(), |space| space.to_string()),
                Some(provenance) => {
                    let mut slices: Vec<_> = provenance.slices.iter().collect();
                    slices.sort_by(|one, other| {
                        (one.object.kind.as_str(), one.object.identity.repr(), one.low, one.high).cmp(&(
                            other.object.kind.as_str(),
                            other.object.identity.repr(),
                            other.low,
                            other.high,
                        ))
                    });
                    slices
                        .into_iter()
                        .map(|part| {
                            format!(
                                "{}[{}:{}:{}/{}]",
                                names.object(&part.object),
                                part.low,
                                part.high,
                                part.stride,
                                part.width
                            )
                        })
                        .collect::<Vec<_>>()
                        .join("|")
                }
            };
            let base = reference.base.map_or_else(String::new, |base| format!("+{}", names.value(base)));
            let segment = reference.segment.map_or_else(String::new, |segment| format!("@{}", names.value(segment)));
            format!("cell({where_}{base}{segment}):{}", reference.width)
        }
        Arg::Symbol(mir::Symbol { space, index, offset, width, addend }) => {
            let displacement = offset + addend;
            let suffix = if displacement > 0 {
                format!("+{displacement}")
            } else if displacement != 0 {
                displacement.to_string()
            } else {
                String::new()
            };
            format!("{}{index}{suffix}:{width}", space.value())
        }
        Arg::FrameAddress(mir::FrameAddress { offset, width, .. }) => {
            let sign = if *offset > 0 { format!("+{offset}") } else { offset.to_string() };
            format!("frame[{sign}]:{width}")
        }
        Arg::Opaque(opaque) => {
            if opaque.name.is_empty() {
                "opaque".to_owned()
            } else {
                opaque.name.clone()
            }
        }
        Arg::FrameSelector(_) => "frameselector".to_owned(),
    }
}

const _INFIX: [mir::Kind; 31] = [
    mir::Kind::Add,
    mir::Kind::Sub,
    mir::Kind::AddCarry,
    mir::Kind::SubBorrow,
    mir::Kind::Mul,
    mir::Kind::Smulhi,
    mir::Kind::Div,
    mir::Kind::Rem,
    mir::Kind::Divmod,
    mir::Kind::Udivmod,
    mir::Kind::And,
    mir::Kind::Or,
    mir::Kind::Xor,
    mir::Kind::Shl,
    mir::Kind::Shr,
    mir::Kind::Sar,
    mir::Kind::Lt,
    mir::Kind::Le,
    mir::Kind::Gt,
    mir::Kind::Ge,
    mir::Kind::Eq,
    mir::Kind::Ne,
    mir::Kind::Below,
    mir::Kind::BelowEq,
    mir::Kind::Above,
    mir::Kind::AboveEq,
    mir::Kind::Fadd,
    mir::Kind::Fsub,
    mir::Kind::Fmul,
    mir::Kind::Fdiv,
    mir::Kind::Fcompare,
];

fn _operation(op: &mir::Op, names: &mut _Names) -> String {
    let args: Vec<String> = op.args.iter().map(|one| _arg(one, names)).collect();
    let results: Vec<String> = op
        .results
        .iter()
        .map(|one| match one {
            Arg::Held(held) => names.value(held.value),
            one => _arg(one, names),
        })
        .collect();
    let defined: Vec<String> = op.defines.iter().map(|one| names.value(*one)).collect();
    let left = if results.is_empty() { defined } else { results.clone() };
    let spelling = if op.op == Some(OpCode::Operation(Operation::FloatUnary)) && !op.name.is_empty() {
        op.name.as_str()
    } else {
        op.kind.as_str()
    };

    if op.kind == mir::Kind::Store && results.len() == 1 && args.len() == 1 {
        return format!("{} <- {}", results[0], args[0]);
    }
    let expression = if _INFIX.contains(&op.kind) && args.len() == 2 {
        format!("{} {spelling} {}", args[0], args[1])
    } else if op.kind == mir::Kind::Call {
        format!("call {}({})", op.name, args.join(", "))
    } else if op.kind == mir::Kind::Branch {
        format!("branch{}", args.first().map_or_else(String::new, |one| format!(" {one}")))
    } else if op.kind == mir::Kind::Return {
        format!("return{}", if args.is_empty() { String::new() } else { format!(" {}", args.join(", ")) })
    } else if args.len() == 1 {
        format!("{spelling} {}", args[0])
    } else if args.is_empty() {
        spelling.to_owned()
    } else {
        format!("{spelling}({})", args.join(", "))
    };
    if left.is_empty() { expression } else { format!("{} <- {expression}", left.join(", ")) }
}

/// Print meaning with incidental block, value, and object identities removed.
pub fn mir_text(lowered: &Lowered) -> String {
    let mut names = _Names::new(&lowered.body);
    let mut lines = vec![format!("function {} entry {}", lowered.name, names.block(lowered.body.entry))];
    for block in &lowered.body.blocks {
        let successors = block.succ.iter().map(|one| names.block(*one)).collect::<Vec<_>>().join(", ");
        let successors = if successors.is_empty() { "-".to_owned() } else { successors };
        lines.push(format!("{} -> {successors}", names.block(block.at)));
        for op in &block.ops {
            let operation = _operation(op, &mut names);
            let target = op.target.map_or_else(String::new, |target| format!(" -> {}", names.block(target)));
            let cases = if op.cases.is_empty() {
                String::new()
            } else {
                let items: Vec<String> =
                    op.cases.iter().map(|(value, at)| format!("({value}, '{}')", names.block(*at))).collect();
                if items.len() == 1 { format!(" ({},)", items[0]) } else { format!(" ({})", items.join(", ")) }
            };
            let floating = match &op.floating {
                None => String::new(),
                Some(floating) => {
                    let inputs = floating.inputs.iter().map(|one| one.as_str()).collect::<Vec<_>>().join(",");
                    format!(
                        " [{inputs}->{};{}/{}]",
                        floating.result.as_str(),
                        floating.precision.as_str(),
                        floating.rounding.as_str()
                    )
                }
            };
            lines.push(format!("  {operation}{target}{cases}{floating}"));
        }
    }
    lines.join("\n") + "\n"
}
